packages/react-components/src/components/TextFieldWithMention/TextFieldWithMention.tsx

```diff
@@ -0,0 +1,629 @@
+              ? 0
+              : Math.min(activeSuggestionIndex + 1, mentionSuggestions.length - 1);
+          setActiveSuggestionIndex(newActiveIndex);
+        }
+      }
+      if (ev.key === 'Enter' && (ev.shiftKey === false || !supportNewline)) {
+        ev.preventDefault();
+        // If we are looking up a mention, select the focused suggestion
+        if (mentionSuggestions.length > 0 && activeSuggestionIndex !== undefined) {
+          const selectedMention = mentionSuggestions[activeSuggestionIndex];
+          if (selectedMention) {
+            onSuggestionSelected(selectedMention);
+            return;
+          }
+        }
+        onEnterKeyDown && onEnterKeyDown();
+      }
+      onKeyDown && onKeyDown(ev);
+    },
+    [onEnterKeyDown, onKeyDown, supportNewline, mentionSuggestions, activeSuggestionIndex, onSuggestionSelected]
+  );
+
+  const debouncedQueryUpdate = useDebouncedCallback(async (query?: string) => {
+    if (query === undefined) {
+      updateMentionSuggestions([]);
+      return;
+    }
+    const suggestions = (await mentionLookupOptions?.onQueryUpdated(query)) ?? [];
+    if (suggestions.length === 0) {
+      setActiveSuggestionIndex(undefined);
+    } else if (activeSuggestionIndex === undefined) {
+      setActiveSuggestionIndex(0);
+    }
+    updateMentionSuggestions(suggestions);
+  }, 500);
+
+  useEffect(() => {
+    return () => {
+      debouncedQueryUpdate.cancel();
+    };
+  }, [debouncedQueryUpdate]);
+
+  const getUpdatedSelectionStartIndexOrEndIndex = useCallback(
+    (
+      inputTextValue: string,
+      tagsValue: TagData[],
+      selectionIndex: number,
+      selectionValue?: number
+    ): number | undefined => {
+      const mentionTag = findMentionTagForSelection(tagsValue, selectionIndex);
+      // don't include boundary cases to show correct selection, otherwise it will show selection at mention boundaries
+      if (
+        mentionTag !== undefined &&
+        mentionTag.plainTextBeginIndex !== undefined &&
+        selectionIndex > mentionTag.plainTextBeginIndex &&
+        selectionIndex < (mentionTag.plainTextEndIndex ?? mentionTag.plainTextBeginIndex)
+      ) {
+        // get updated selection index
+        const newSelectionIndex = findNewSelectionIndexForMention({
+          tag: mentionTag,
+          textValue: inputTextValue,
+          currentSelectionIndex: selectionIndex,
+          previousSelectionIndex: selectionValue ?? inputTextValue.length
+        });
+        return newSelectionIndex;
+      }
+      return undefined;
+    },
+    []
+  );
+
+  // Update selections index in mention to navigate by words
+  const updateSelectionIndexesWithMentionIfNeeded = useCallback(
+    (
+      event: FormEvent<HTMLInputElement | HTMLTextAreaElement>,
+      inputTextValue: string,
+      selectionStartValue: number | undefined,
+      selectionEndValue: number | undefined,
+      tagsValue: TagData[]
+    ): void => {
+      const currentSelectionStart =
```

Comment:
This main part of the function could be separated out as a single const function outside of the component body 

The purpose of this function is to 
1. calculate updatedStartIndex and updatedEndIndex 
2. Set those 2 values into the state

Only 2nd part is this component related while the main part for 1st purpose could be extract as a single function (util function maybe?)

The main purpose here is to make component body as minimized as possible